### PR TITLE
feat(connector): [Stax] Implement Bank Debits and Webhooks for Connector Stax

### PR DESCRIPTION
**Pull request by deepanshu-iiitu**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
Stax is a low cost payment connector from North America.
Add Bank debits and Web-hooks support for Stax.

Supported Flows:

- Authorize
- Capture
- Sync
- Refund
- Void
- Web-hooks

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context
Add Webhooks and bankdebits for Payments and Refunds flow. 

## How did you test it?
Testing was carried out using Postman Calls
Screenshot 2023-08-08 at 7 39 27 PM
`{
    "amount": 5,
    "currency": "USD",
    "confirm": true,
    "capture_method": "automatic",
    "customer_id": "cust143",
    "email": "customer143@juspay.in",
    "payment_method": "bank_debit",
    "payment_method_type": "ach",
    "payment_method_data": {
        "bank_debit": {
            "ach_bank_debit": {
                "billing_details": {
                    "name": "Deepanshu Bansal",
                    "email": "customer139@juspay.in"
                },
                "account_number": "9876543210",
                "routing_number": "021000021",
                "card_holder_name": "Steven Smith",
                "bank_account_holder_name": "Steven Smith",
                "bank_name": "bank_of_america",
                "bank_type": "savings",
                "bank_holder_type": "personal"
            }
        }
    },
    "business_label": "default",
    "business_country": "US"
}`

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code
- [x] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/connector/stax/transformers.rs [hunk 1]
```diff
@@ -138,6 +163,28 @@ impl TryFrom<&types::TokenizationRouterData> for StaxTokenRequest {
...
+                    ..
+                },
+            ) => {
+                let stax_bank_data = StaxBankTokenizeData {
+                    person_name: billing_details.name,
+                    bank_account: account_number,
+                    bank_routing: routing_number,
+                    bank_name: bank_name.unwrap_or(Secret::new("".to_string())),
+                    bank_type: bank_type.unwrap_or(Secret::new("".to_string())),
+                    bank_holder_type: bank_holder_type.unwrap_or(Secret::new("".to_string())),
```
Comment (ArjunKarthik) [inline crates/router/src/connector/stax/transformers.rs:183]: In case of None we should throw missing field error

**Review by ArjunKarthik — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/stax.rs [hunk 2]
```diff
@@ -760,22 +762,63 @@ impl ConnectorIntegration<api::RSync, types::RefundsData, types::RefundsResponse
...
+            .parse_struct("StaxWebhookBody")
+            .change_context(errors::ConnectorError::WebhookReferenceIdNotFound)?;
+
+        match webhook_body.transaction_type {
+            stax::StaxWebhookEventType::Refund => {
+                Ok(api_models::webhooks::ObjectReferenceId::RefundId(
+                    api_models::webhooks::RefundIdType::ConnectorRefundId(webhook_body.id),
+                ))
+            }
+            _ => Ok(api_models::webhooks::ObjectReferenceId::PaymentId(
```
[the remark this answers is not in the document]
Comment (deepanshu-iiitu) [reply crates/router/src/connector/stax.rs:786]: This case is now handled explicitly.

File: crates/router/src/connector/stax/transformers.rs [hunk 3]
```diff
@@ -354,3 +401,24 @@ impl TryFrom<types::RefundsResponseRouterData<api::RSync, RefundResponse>>
...
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "snake_case")]
+pub enum StaxWebhookEventType {
+    PreAuth,
+    Capture,
+    Charge,
+    Void,
+    Refund,
+    #[serde(other)]
+    Unknown,
```
[the remark this answers is not in the document]
Comment (deepanshu-iiitu) [reply crates/router/src/connector/stax/transformers.rs:427]: As discussed, this is not changed

File: crates/common_utils/src/crypto.rs [hunk 4]
```diff
@@ -130,7 +130,7 @@ impl VerifySignature for NoAlgorithm {
         _signature: &[u8],
         _msg: &[u8],
     ) -> CustomResult<bool, errors::CryptoError> {
-        Ok(true)
+        Ok(false)
```
[the remark this answers is not in the document]
Comment (deepanshu-iiitu) [reply crates/common_utils/src/crypto.rs]: Done

File: crates/api_models/src/payments.rs [hunk 5]
```diff
@@ -672,6 +672,15 @@ pub enum BankDebitData {
 
         #[schema(value_type = String, example = "John Doe")]
         bank_account_holder_name: Option<Secret<String>>,
+
+        #[schema(value_type = String, example = "ACH")]
+        bank_name: Option<Secret<String>>,
```
[the remark this answers is not in the document]
Comment (deepanshu-iiitu) [reply crates/api_models/src/payments.rs]: Done

File: crates/router/src/connector/stax/transformers.rs (hunk 1 above)
Comment (deepanshu-iiitu) [reply crates/router/src/connector/stax/transformers.rs]: Done

File: crates/api_models/src/payments.rs [hunk 6]
```diff
@@ -672,6 +672,15 @@ pub enum BankDebitData {
 
         #[schema(value_type = String, example = "John Doe")]
         bank_account_holder_name: Option<Secret<String>>,
+
+        #[schema(value_type = String, example = "ACH")]
+        bank_name: Option<enums::BankNames>,
+
+        #[schema(value_type = String, example = "Checking")]
+        bank_type: Option<Secret<String>>,
```
Comment (ArjunKarthik) [inline crates/api_models/src/payments.rs:680]: Let's change this to enum

File: crates/api_models/src/payments.rs [hunk 7]
```diff
@@ -672,6 +672,15 @@ pub enum BankDebitData {
...
+
+        #[schema(value_type = String, example = "Personal")]
+        bank_holder_type: Option<Secret<String>>,
```
Comment (ArjunKarthik) [inline crates/api_models/src/payments.rs:683]: Same here

File: crates/router/src/connector/stax.rs [hunk 8]
```diff
@@ -751,24 +754,79 @@ impl ConnectorIntegration<api::RSync, types::RefundsData, types::RefundsResponse
...
+        match webhook_body.transaction_type {
+            stax::StaxWebhookEventType::Refund => {
+                Ok(api_models::webhooks::ObjectReferenceId::RefundId(
+                    api_models::webhooks::RefundIdType::ConnectorRefundId(webhook_body.id),
+                ))
+            }
+            stax::StaxWebhookEventType::Unknown => {
+                Err(errors::ConnectorError::WebhookEventTypeNotFound.into())
+            }
+            _ => Ok(api_models::webhooks::ObjectReferenceId::PaymentId(
```
Comment (ArjunKarthik) [inline crates/router/src/connector/stax.rs:786]: Please remove default case

File: crates/router/src/connector/stax.rs [hunk 9]
```diff
@@ -751,24 +754,79 @@ impl ConnectorIntegration<api::RSync, types::RefundsData, types::RefundsResponse
...
+                true => api::IncomingWebhookEvent::RefundSuccess,
+                false => api::IncomingWebhookEvent::RefundFailure,
+            },
+            StaxWebhookEventType::Capture | StaxWebhookEventType::Charge => {
+                match &details.success {
+                    true => api::IncomingWebhookEvent::PaymentIntentSuccess,
+                    false => api::IncomingWebhookEvent::PaymentIntentFailure,
+                }
+            }
+            _ => api::IncomingWebhookEvent::EventNotSupported,
```
Comment (ArjunKarthik) [inline crates/router/src/connector/stax.rs:819]: Please avoid using default case

File: crates/router/src/connector/stax/transformers.rs [hunk 10]
```diff
@@ -37,6 +38,19 @@ impl TryFrom<&types::PaymentsAuthorizeRouterData> for StaxPaymentsRequest {
                     payment_method_id: Secret::new(item.get_payment_method_token()?),
                 })
             }
+            api::PaymentMethodData::BankDebit(_) => {
+                let mut pre_auth = false;
+                if let Some(enums::CaptureMethod::Manual) = item.request.capture_method {
```
Comment (ArjunKarthik) [inline crates/router/src/connector/stax/transformers.rs:43]: You can make use of `is_auto_capture` from utils

File: crates/router/src/connector/stax/transformers.rs [hunk 11]
```diff
@@ -138,6 +164,40 @@ impl TryFrom<&types::TokenizationRouterData> for StaxTokenRequest {
...
+                    bank_holder_type,
+                    ..
+                },
+            ) => {
+                let stax_bank_data = StaxBankTokenizeData {
+                    person_name: billing_details.name,
+                    bank_account: account_number,
+                    bank_routing: routing_number,
+                    bank_name: bank_name.get_required_value("bank_name").change_context(
+                        errors::ConnectorError::MissingRequiredField {
```
Comment (ArjunKarthik) [inline crates/router/src/connector/stax/transformers.rs:183]: Use `missing_field_err` from the utils

File: crates/router/src/connector/stax/transformers.rs [hunk 12]
```diff
@@ -138,6 +164,40 @@ impl TryFrom<&types::TokenizationRouterData> for StaxTokenRequest {
...
+                            field_name: "bank_name",
+                        },
+                    )?,
+                    bank_type: bank_type.get_required_value("bank_type").change_context(
+                        errors::ConnectorError::MissingRequiredField {
```
Comment (ArjunKarthik) [inline crates/router/src/connector/stax/transformers.rs:188]: Use `missing_field_err` from the utils

File: crates/router/src/connector/stax/transformers.rs [hunk 13]
```diff
@@ -138,6 +164,40 @@ impl TryFrom<&types::TokenizationRouterData> for StaxTokenRequest {
...
+                            field_name: "bank_type",
+                        },
+                    )?,
+                    bank_holder_type: bank_holder_type
+                        .get_required_value("bank_holder_type")
+                        .change_context(errors::ConnectorError::MissingRequiredField {
```
Comment (ArjunKarthik) [inline crates/router/src/connector/stax/transformers.rs:194]: Use `missing_field_err` from the utils

**Review by ArjunKarthik — CHANGES_REQUESTED** [review]

File: crates/api_models/src/payments.rs (hunk 6 above)
Comment (deepanshu-iiitu) [reply crates/api_models/src/payments.rs]: Done

File: crates/router/src/connector/stax.rs (hunk 8 above)
Comment (deepanshu-iiitu) [reply crates/router/src/connector/stax.rs]: Removed

File: crates/router/src/connector/stax.rs (hunk 9 above)
Comment (deepanshu-iiitu) [reply crates/router/src/connector/stax.rs]: Done

File: crates/router/src/connector/stax/transformers.rs (hunk 10 above)
Comment (deepanshu-iiitu) [reply crates/router/src/connector/stax/transformers.rs]: Done

File: crates/router/src/connector/stax/transformers.rs (hunk 11 above)
Comment (deepanshu-iiitu) [reply crates/router/src/connector/stax/transformers.rs]: Done

File: crates/router/src/connector/stax/transformers.rs (hunk 12 above)
Comment (deepanshu-iiitu) [reply crates/router/src/connector/stax/transformers.rs]: Done

File: crates/router/src/connector/stax/transformers.rs (hunk 13 above)
Comment (deepanshu-iiitu) [reply crates/router/src/connector/stax/transformers.rs]: Done

**Review by ArjunKarthik — APPROVED** [review]

**Review by ashokkjag — APPROVED** [review]

**Review by ArjunKarthik — APPROVED** [review]

**Pull request merged**